### ISSUE TITLE
Twisted Web server does not accept chunked encoded request bodies

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3385]
|<img alt="Quartz25's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Quartz25 reported|
|-|-|
|Trac ID|trac#3385|
|Type|defect|
|Created|2008-08-08 21:52:47Z|
|Branch|https://github.com/twisted/twisted/tree/server-receive-chunked-3385|

As documented in [this bug report](https://bugs.launchpad.net/zfilenet/+bug/253117), Twisted Web does not support the "chunked" transfer encoding for HTTP, a requirement in the HTTP/1.1 specification.

After fixing this in my own program, I have written a patch for use in the general Twisted library, along with a unit test.

Attachments:

* [patch-3385.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7d8/7d8c2147550af2e791898135ec1920eb7c29db74/4968676be34cc314f3d7576b9ea8b44d2b792e7b.diff) (3601 bytes) - added by Quartz25 on 2008-08-08 21:53:37Z - Implementation of Twisted Web chunked encoding
* [patch-3385.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7d8/7d8c2147550af2e791898135ec1920eb7c29db74/b090f644a641e04b74a04f8411776da43f1c5fb0.diff) (4805 bytes) - added by Quartz25 on 2008-08-18 01:02:54Z - Revised patch for chunked encoding

<details><summary>Searchable metadata</summary>

```
trac-id__3385 3385
type__defect defect
reporter__Quartz25 Quartz25
priority__highest highest
milestone__ 
branch__branches_server_receive_chunked_3385 branches/server-receive-chunked-3385
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1218232367000000 1218232367000000
changetime__1221057295000000 1221057295000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to Quartz25|
|-|-|

Thanks for the patch.  A few comments:

  * The new `chunked` and `chunk_buffer` attributes of `HTTPChannel` should be documented.  They should probably also be private.  `chunk_buffer` should be named like `chunkBuffer`, as well, to conform to the naming standard.
  * There's no unit test coverage for `chunked` and `chunk_buffer` being reset to their initial values in `allContentReceived`.
  * Using repeated string concatenation to store the chunks the entire chunk is received is inefficient.  A StringIO would probably be the best option.
  * Instead of setting length to 100000 to force use of a file for content storage, it'd probably be pretty reasonable to check to see if length is `None` in `gotLength` and make the decision there.

Slightly bigger picture perspective: it would probably simplify the code and make it easier to write good unit tests if different transfer-encodings (including even the default, perhaps) were implemented in different classes and an instance of the appropriate class were created and used by `HTTPChannel` to interpret data received in `rawDataReceived`.  This will make it much simpler to support other transfer-encodings as well.  If you're interested in making this refactoring, that'd be great, otherwise I may give it a try (after the above points are addressed).

**Comment by twisted-trac** [comment]
|<img alt="Quartz25's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Quartz25 commented|
|-|-|

Replying to [exarkun](#note_1):

I'm now working on revising the patch.  Sadly, I won't be able to help with a refactor, but I do think it's a good idea.

>   * The new `chunked` and `chunk_buffer` attributes of `HTTPChannel` should be documented.  They should probably also be private.  `chunk_buffer` should be named like `chunkBuffer`, as well, to conform to the naming standard.

Where do I document them?  Should I use docstrings?  I don't see any for other HTTPChannel attributes.

Also, I think chunked might be useful to be public.  I'll cite the example of "t.w.http.Request.chunked" as being public (yet not documented).  The chunk buffer, on the other hand, I agree with you probably should be private.

>   * Using repeated string concatenation to store the chunks the entire chunk is received is inefficient.  A StringIO would probably be the best option.

The chunk buffer doesn't store all the chunks.  It only stores the unparsed data until a chunk is completely sent, then it gets sent to the Request (which ends up using a temporary file).  If I used StringIO, I would have to append the data, then once a chunk is parsed successfully, clear the whole StringIO object and rewrite all the data after the chunk.  It seems like that would be less efficient.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> Where do I document them? Should I use docstrings? I don't see any for other HTTPChannel attributes.

Yes, in the class docstring.  The code is old, so it doesn't all live up to the current standards.

> The chunk buffer doesn't store all the chunks. It only stores the unparsed data until a chunk is completely sent, then it gets sent to the Request (which ends up using a temporary file). If I used StringIO, I would have to append the data, then once a chunk is parsed successfully, clear the whole StringIO object and rewrite all the data after the chunk. It seems like that would be less efficient. 

Sorry, I noticed it was just one chunk, but failed to properly edit my comment before submitting.  Even for one chunk at a time, a StringIO will be better (and, unfortunately, a cStringIO will be best).  Repeated string concatenation is quadratic, whereas writing to a StringIO is linear (and writing to a cStringIO has memory benefits).  It'd be best to have a benchmark to demonstrate that it's a clear win in all cases for this code, but even without one I'm pretty confident saying that this is how it should be done.  At the _very_ least, it should be a list that gets appended to - StringIO is only a slight win over that, and possibly not even in all cases.

**Comment by twisted-trac** [comment]
|<img alt="Quartz25's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Quartz25 commented|
|-|-|

I've made the requested changes to the patch, and uploaded it as patch-3385.2.diff.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [24629]) Branching to 'server-receive-chunked-3385'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks!  The patch looks great.  I'm going to try to implement the refactoring I described earlier.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [24630]) Apply patch-3385.2.diff

refs [#3385](https://github.com/twisted/twisted/issues/3385)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|


This looks pretty awesome.

  1. `ChunkedTransferEncoding`'s docstring should probably mention that it's meant to be used as a 'transfer decoder' protocol.  i.e. it should be made clear that it does not parse HTTP with chunked transfer encoding.  The right place to say this might be a docstring for `dataReceived`, saying what it expects `data` to be.
  2. I have been leaning towards making things private unless they really need to be otherwise.  `ChunkedTransferEncoding` looks like it is a candidate to be private - is there any reason that an application would want to construct one of these itself?  The 'transfer decoder' mechanism looks pretty cool, though.  Perhaps we could anticipate more of these and make a `twisted.web._encodings` module, where `ChunkedTransferEncoding` could live?  That might also retard http.py's inevitable march to 2k lines.
  3. `chunkCallback` is a bit of a misleading name; it's called each time data is received, not for each chunk.
    1. ... each time ... `data` ... is `Received`.  And there's a `finishCallback` as well for when this virtual `connection` is `Lost`.  Smells to me that this is a wrapper protocol, and the thing it's wrapping is another protocol.
    2. This also implies that `connectionMade` and `connectionLost` ought to be relayed to `ChunkedTransferEncoding` so that it can propagate them to the wrapped class.  Receipt of the final 'close' chunk would make it possible to tell whether to use CONNECTION_LOST or CONNECTION_DONE, too.
    3. I want to say something about `Content-Encoding` but I don't know what.  I guess maybe it's not actually relevant but it seems like it will be really simple to implement using a similar (or the same) mechanism.
  4. This could really benefit from a more explicit state machine.  `dataReceived` is a little too long for my taste.
  5. Why a `finish` flag rather than a `last-trailer` state?

Aside from the issue of not making it public too soon, these issues could all be deferred until future tickets if there's anything hard; documentation is good, test coverage sufficient, they pass.  The implementation looks pretty good, and this is a good feature to have.  As far as I'm concerned it could be merged with the addition of a single "`_`" - and without even that, if there's a reason that this should be a public thing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

All changes in r24663

  1. Added docs to the class docstring and gave dataReceived a docstring that repeats the point
  1. Made it private.  Didn't move it - if we write a second one, maybe we can think about this, and since it's private it'll be trivial.  I'd be more inclined to do this now, but I think there are only two Transfer-Encodings, and only one that anyone actually uses.
  1. I renamed it dataCallback.  Your point about the rest of the `IProtocol` interface is well taken, but I opted to make a note of these points in a comment and make `_ChunkedTransferEncoding` not be a `Protocol` at all for now.  Content-Encoding may benefit from a similar solution, I guess?  But I think it is indeed a separate task.
  1. I could make it one, if you really want.  But then it'd be a lot slower, right?  It does fit on one page, fwiw.
  1. I wanted to share code and using an unfortunate flag seemed to be the best way to do that.  Any suggestions?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

  1. OK
  2. Right, that was more for content-encodings than content-transfer-encodings.  I like the idea of using `IProtocol`, but the real benefit involves exposing some method on `Request` to shove an application-level one in there.  We can burn that bridge when we come to it, which I hope will be soon.
  3. Given that the class is now private and the comments are for maintainers, I think they should be in the class docstring rather than a comment.
  4. Eh, it's fine.
  5. In the spirit of 4; it's fine.  Nothing comes to mind immediately, adding an extra state would probably just make it longer.

Go ahead and merge, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [24665]) turn comment into docstring

refs [#3385](https://github.com/twisted/twisted/issues/3385)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [24667]) Merge server-receive-chunked-3385 

Author: exarkun, Quartz25
Reviewer: glyph
Fixes: [#3385](https://github.com/twisted/twisted/issues/3385)

Add a general `Transfer-Encoding: chunked` parser to Twisted Web and use it
in the HTTP server protocol implementation to accept request bodies which
use that encoding.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `reopened`|
|-|-|

(In [24669]) Revert r24667 - API doc failure

Reopens: [#3385](https://github.com/twisted/twisted/issues/3385)

The new `twisted.web.http.Request.gotLength` docstring is not valid epytext.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [24670]) fields must be the last thing in a docstring

refs [#3385](https://github.com/twisted/twisted/issues/3385)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Additionally, this failure occurs on Windows:

```
[FAIL]: twisted.web.test.test_http.ParsingTestCase.test_chunkedEncoding

Traceback (most recent call last):
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\test\test_http.py", line 652, in test_chunkedEncoding
    self.runRequest(httpRequest, MyRequest)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\test\test_http.py", line 408, in runRequest
    a.dataReceived(byte)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\protocols\basic.py", line 239, in dataReceived
    return self.rawDataReceived(data)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\http.py", line 1393, in rawDataReceived
    self._transferDecoder.dataReceived(data)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\http.py", line 1228, in dataReceived
    self.finishCallback(data)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\http.py", line 1333, in _finishRequestBody
    self.allContentReceived()
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\http.py", line 1389, in allContentReceived
    req.requestReceived(command, path, version)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\http.py", line 712, in requestReceived
    self.process()
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\test\test_http.py", line 643, in process
    testcase.assertIsInstance(self.content, file)
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\trial\unittest.py", line 414, in failUnlessIsInstance
    self.fail("%r is not an instance of %s" % (instance, classOrTuple))
twisted.trial.unittest.FailTest: <open file '<fdopen>', mode 'w+b' at 0x066CB020> is not an instance of <type 'file'>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [24712]) Check fileno instead of the type

refs [#3385](https://github.com/twisted/twisted/issues/3385)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Build results:

 http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/server-receive-chunked-3385

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve set owner to @exarkun|
|-|-|

There is an extra blank line before test_extensions. I'm not totally happy about the dataReceived method either (I have a small screen, I have to scroll :)), but that's not major. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [24768]) Remove extraneous blank line

refs [#3385](https://github.com/twisted/twisted/issues/3385)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun set status to `closed`|
|-|-|

(In [24769]) Merge server-receive-chunked-3385

Author: Quartz25, exarkun
Reviewer: exarkun, glyph, therve
Fixes: [#3385](https://github.com/twisted/twisted/issues/3385)

Add support for chunked transfer-encoded request bodies to
`twisted.web.http.HTTPChannel`.  Chunked request bodies will
be delivered to application code just as non-chunked bodies
would be.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
